perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/PerunBlImpl.java

```diff
@@ -153,6 +171,34 @@ public PerunSession getPerunSession(PerunPrincipal principal, PerunClient client
 		return perunSession;
 	}
 
+	/**
+	 * Iteratively searches through additional identifiers trying to find UES with the same identifier.
+	 * Returns first found UES or null when no matching UES found.
+	 *
+	 * @param principal PerunPrincipal which contains additionalIdentifiers
+	 * @param internalSession PerunSession to retrieve UserExtSource
+	 * @return UserExtSource found using additionalIdentifiers or null
+	 * @throws InternalErrorException When attribute is not defined in Perun
+	 */
+	public UserExtSource getUserExtSourceFromMultipleIdentifiers(PerunPrincipal principal, PerunSession internalSession) throws InternalErrorException {
+		String additionalIdentifiers = principal.getAdditionalInformations().get(additionalIdentifiersAttributeName);
+		UserExtSource ues = null;
+		for(String identifier : additionalIdentifiers.split(multivalueAttributeSeparatorRegExp)) {
```

Comment:
You can get NullPointerException if additionalIdentifiers will be null.

Reply:
Can that actually happen with the check on lines 139 or 150? Not sure but I expected this check will prevent the attribute to be null.

Reply:
This is separate public method. If anybody calls it, it's just about returned value of the call "principal.getAdditionalInformations().get(additionalIdentifiersAttributeName)". So if you are sure this call will never return null, than you don't need to check it. Otherwise you should.

Reply:
Makes total sense. How to handle it? Log and throw AttributeNotFoundException?

Reply:
The question is what does it mean:
- it is unexpected event? Than InternalErrorException is good choice
- It is because there is something missing which should never be missing? Then any error with more specific info is good like - 'CoreAttributeNotFoundException' or something like that (be careful about simple word "Attribute" because it is overloaded in Perun)
- it is normal situation? then you can log it (or not) and return empty ues (null in this case)

Reply:
Used UserExtSourceNotExistsException as found in UsersManagerImpl.java on line 573.